nucliadb/nucliadb/common/maindb/tikv.py

```diff
@@ -218,21 +213,135 @@ async def count(self, match: str) -> int:
+    async def get(self, key: str) -> Optional[bytes]:
+        return await self.data_layer.get(key)
+
+    async def set(self, key: str, value: bytes) -> None:
+        return await self.data_layer.set(key, value)
+
+    async def delete(self, key: str) -> None:
+        return await self.data_layer.delete(key)
+
+    async def keys(
+        self, match: str, count: int = DEFAULT_SCAN_LIMIT, include_start: bool = True
+    ):
+        # XXX must have connection outside of current txn
+        conn_holder = self.driver.get_connection_holder()
+        txn = await conn_holder.get_snapshot()
+        dl = TiKVDataLayer(txn)
+
+        async for key in dl.keys(match, count, include_start):
+            yield key
+
+    async def count(self, match: str) -> int:
+        return await self.data_layer.count(match)
+
+
+class ReadOnlyTiKVTransaction(Transaction):
+    driver: TiKVDriver
+
+    def __init__(self, connection: asynchronous.Snapshot, driver: TiKVDriver):
+        self.connection = connection
+        self.data_layer = TiKVDataLayer(connection)
+        self.driver = driver
+
+    async def abort(self):
+        ...
+
+    async def commit(self):
+        raise Exception("Cannot commit transaction in read only mode")
+
+    async def batch_get(self, keys: list[str]) -> list[Optional[bytes]]:
+        return await self.data_layer.batch_get(keys)
+
+    async def get(self, key: str) -> Optional[bytes]:
+        return await self.data_layer.get(key)
+
+    async def set(self, key: str, value: bytes) -> None:
+        raise Exception("Cannot set in read only transaction")
+
+    async def delete(self, key: str) -> None:
+        raise Exception("Cannot delete in read only transaction")
+
+    async def keys(
+        self, match: str, count: int = DEFAULT_SCAN_LIMIT, include_start: bool = True
+    ):
+        async for key in self.data_layer.keys(match, count, include_start):
+            yield key
+
+    async def count(self, match: str) -> int:
+        return await self.data_layer.count(match)
+
+
+class ConnectionHolder:
+    _txn_connection: asynchronous.TransactionClient
 
     def __init__(self, url: list[str]):
-        if TiKV is False:
-            raise ImportError("TiKV is not installed")
         self.url = url
         self.connect_lock = asyncio.Lock()
         self.connect_in_progress = False
 
-    async def initialize(self):
-        async with self.connect_lock:
-            if self.initialized is False and self.tikv is None:
-                self.tikv = await TransactionClient.connect(self.url)
-            self.initialized = True
+    async def initialize(self) -> None:
+        self._txn_connection = await asynchronous.TransactionClient.connect(self.url)
+
+    async def get_snapshot(
+        self, timestamp: Optional[float] = None, retried: bool = False
+    ) -> asynchronous.Snapshot:
+        if self.connect_in_progress:  # pragma: no cover
+            async with self.connect_lock:
```

Comment:
are both `connect_in_progress` flag and the lock needed? Or could we use just the lock?